makeflow/src/makeflow_wrapper_enforcement.c

```diff
@@ -0,0 +1,140 @@
+	w->command = xxstrdup(enforcer_pattern "%%");
+}
+
+char *makeflow_wrap_enforcer( char *result, struct dag_node *n, struct makeflow_wrapper *w, struct list *input_list, struct list *output_list )
+{
+	if(!w) return result;
+
+	struct list *enforcer_paths = list_create();
+	struct dag_file *f;
+	FILE *enforcer;
+	char *enforcer_path = string_format(enforcer_pattern "%d", n->nodeid);
+	char *mountlist_path = string_format(mountlist_pattern "%d", n->nodeid);
+	char *tmp_path = string_format(tmp_pattern "%d", n->nodeid);
+	char *vartmp_path = string_format(vartmp_pattern "%d", n->nodeid);
+	char *shm_path = string_format(shm_pattern "%d", n->nodeid);
+
+	/* Create local directories to bind to /tmp, /var/tmp, and /dev/shm */
+	if (mkdir(tmp_path, S_IRWXU) || mkdir(vartmp_path, S_IRWXU) || mkdir(shm_path, S_IRWXU)) {
+		fatal("could not create temp directory: %s", strerror(errno));
+	}
+
+	list_push_tail(enforcer_paths, dag_file_lookup_or_create(n->d, tmp_path));
+	list_push_tail(enforcer_paths, dag_file_lookup_or_create(n->d, vartmp_path));
+	list_push_tail(enforcer_paths, dag_file_lookup_or_create(n->d, shm_path));
+
+	/* make an invalid mountfile to send */
+	int mountlist_fd = open(mountlist_path, O_WRONLY|O_CREAT|O_EXCL, S_IRUSR|S_IWUSR);
+	if (mountlist_fd == -1) {
+		fatal("could not create `%s': %s", mountlist_path, strerror(errno));
+	}
+	write(mountlist_fd, "mountlist\n", 10);
+	close(mountlist_fd);
+
+	list_push_tail(enforcer_paths, dag_file_lookup_or_create(n->d, mountlist_path));
+
+	/* and generate a wrapper script with the current nodeid */
+	int enforcer_fd = open(enforcer_path, O_WRONLY|O_CREAT|O_EXCL, S_IRWXU);
+	if (enforcer_fd == -1 || (enforcer = fdopen(enforcer_fd, "w")) == NULL) {
+		fatal("could not create `%s': %s", enforcer_path, strerror(errno));
+	}
+	fchmod(enforcer_fd, 0755);
+	fprintf(enforcer, "#!/bin/sh\n\n");
+	fprintf(enforcer, "MOUNTFILE='%s'\n", mountlist_path);
+	fprintf(enforcer, "cat > \"$MOUNTFILE\" <<EOF\n");
+	fprintf(enforcer, "/\t\trx\n");
+	fprintf(enforcer, "/tmp\t$PWD/%s\trwx\n", tmp_path);
+	fprintf(enforcer, "/var/tmp\t$PWD/%s\trwx\n", vartmp_path);
+	fprintf(enforcer, "/dev/shm\t$PWD/%s\trwx\n", shm_path);
+	fprintf(enforcer, "/dev/null\trwx\n");
+	fprintf(enforcer, "/dev/zero\trwx\n");
+	fprintf(enforcer, "/dev/full\trwx\n");
+	fprintf(enforcer, "/dev/random\trwx\n");
+	fprintf(enforcer, "/dev/urandom\trwx\n");
+	fprintf(enforcer, "/home\t\tDENY\n");
+	list_first_item(input_list);
+	while((f=list_next_item(input_list))) {
+		fprintf(enforcer, "$PWD/%s\trwx\n", f->filename);
+	}
+	list_first_item(output_list);
+	while((f=list_next_item(output_list))) {
+		fprintf(enforcer, "$PWD/%s\trwx\n", f->filename);
+	}
+	fprintf(enforcer, "EOF\n\n");
+	fprintf(enforcer, "./parrot_run -m \"$MOUNTFILE\" -- \"$@\"\n");
+	fprintf(enforcer, "exit $?\n");
+	fclose(enforcer);
+
+	list_push_tail(enforcer_paths, dag_file_lookup_or_create(n->d, enforcer_path));
+
+	makeflow_log_file_expectation(n->d, enforcer_paths);
```

Comment:
To be more clear, the enforcer_paths are inputs that are created before the node is run. Following a node's completion we move outputs from expected to existing, and inputs from existing to completed if they are no longer needed. The issue is that if the files aren't logged prior to the node's completion we treat the inputs as user provided and don't delete them. Wrapper outputs already get logged as expected->existing, but the inputs are not because a node does not explicitly create it.